### DISCUSSION TITLE
Confusing about unconsistenly unref behavior of nested setTimout

### DISCUSSION THREAD

**Discussion by conanjs** [nodejs/help#4654] [Q&A]
Hi, I have a test as code below but don't understand its behavior:
The output below will sometimes log "1 2 3", but sometimes just log "1". It is not consistent.

- As I understand it, when I use the "unref" method of a Timeout, it won't require the Node Event Loop to remain active.
- I'm quite not sure, but I assume that NodeJS has a check at the final phase before beginning a new tick. And because the Timer phase is the first phase of the macro task.
- And one more thing is nested setTimeout callback won't block the event loop because it will defer newly nested callback until the next tick of the event loop

So I think when I called unref method for both 'x' and 'y'. Before beginning a new tick, there is no timeout reference, and event loop should exit. So what I expected from the below code is just logging 1.

but actually, it is not consistent as I described above.

Can anyone help me explain its behavior, I'm a person who really curious, and interested in some stuff like this.

Thank you!

```
setTimeout(() => {
    console.log("1");

    const x = setTimeout(() => {
        console.log("2");
    });
    const y = setTimeout(() => {
        console.log("3");
    });

    const start = Date.now();
    while (Date.now() - start <= 2000);

    x.unref();
    y.unref();
});
```

**Comment by liudonghua123** [comment]
I tested with the latest node 22.0.0 a few times. It always only log 1 and about 2.6s then it exits.

```
Liu.D.H  playground   2.663s  14:21 > node timeout_unref.js
1

Liu.D.H  playground   2.63s  14:22 > node timeout_unref.js
1

Liu.D.H  playground   2.084s  14:22 > node -v
v22.0.0

Liu.D.H  playground   47ms  14:22 > node timeout_unref.js
1

Liu.D.H  playground   2.09s  14:22 >
```

**Reply by conanjs** [reply]
Yes, as I tested with 22.0.0 like you, it looks like works normally and as expected. I also use Bun runtime to test. it works the same no matter how many times run it. It just logs 1 (expected)

Node version <=22.12.2 has this problem.
